### PR TITLE
Glir logger

### PR DESCRIPTION
**Pull request by izaid**
This PR implements a Glir logger that logs all Glir commands to a specified file. It is run via "python vispy --vispy-glir-file=somefile.txt".

### CODE REVIEW

File: vispy/__init__.py [hunk 1]
```diff
@@ -27,6 +27,13 @@
 # Nice string for the version (mimic how IPython composes its version str)
 __version__ = '-'.join(map(str, version_info)).replace('-', '.', 2).strip('-')
 
+from argparse import ArgumentParser
+
+parser = ArgumentParser()
+parser.add_argument('--vispy-glir-file', help='Export glir commands to specified file.')
+args = parser.parse_args()
+glir_file = args.vispy_glir_file
+
```
Comment (rossant) [inline vispy/__init__.py]: Should we use the system in https://github.com/vispy/vispy/blob/5b01479e829e8f4b2dd348dff51ce864c1f834a5/vispy/util/config.py#L74 instead?

Comment (izaid) [inline vispy/__init__.py]: Yes, absolutely! Didn't know about that.

File: vispy/util/logs.py [hunk 2]
```diff
@@ -327,3 +330,14 @@ def _handle_exception(ignore_callback_errors, print_callback_errors, obj,
             else:  # == 'node':
                 logger.error("Drawing node %s repeat %s"
                              % (node, this_print))
+
+class NumPyJSONEncoder(json.JSONEncoder):
+    def default(self, obj):
+        from ..app.backends._ipynb_util import _serialize_buffer
```
Comment (rossant) [inline vispy/util/logs.py]: move this function to vispy util?

File: vispy/util/logs.py [hunk 3]
```diff
@@ -327,3 +330,14 @@ def _handle_exception(ignore_callback_errors, print_callback_errors, obj,
...
+
+        if isinstance(obj, np.ndarray):
+            return _serialize_buffer(obj, array_serialization='base64')
+        elif isinstance(obj, np.generic):
+            return obj.item()
+
+        return json.JSONEncoder.default(self, obj)
```
Comment (rossant) [inline vispy/util/logs.py]: new line at end of file

**Comment by rossant** [comment]
The approach LGTM. Can you add some tests?

File: vispy/util/logs.py (hunk 2 above)
Comment (izaid) [inline vispy/util/logs.py]: Where in vispy.util do you want it? It is currently in vispy.util.logs.

Comment (rossant) [inline vispy/util/logs.py]: sorry I was talking about _serialize_buffer

Comment (izaid) [inline vispy/util/logs.py]: Yes, I'll move that. What about _serialize_command and _serialize_item? Should I be using those also?

Comment (rossant) [inline vispy/util/logs.py]: if _serialize_buffer depends on them yes

Comment (izaid) [inline vispy/util/logs.py]: _serialize_buffer doesn't depend on them. Was your intention with those to use them for glir commands?

Comment (rossant) [inline vispy/util/logs.py]: OK you can just move _serialize_buffer and leave the others where they are

File: vispy/gloo/glir.py [hunk 4]
```diff
@@ -294,7 +294,75 @@ def convert_shaders(self):
...
+                ob.set_framebuffer(*args)
+            elif cmd == 'SHADERS':  # Program
+                ob.set_shaders(*args)
+            elif cmd == 'WRAPPING':  # Texture1D, Texture2D, Texture3D
+                ob.set_wrapping(*args)
+            elif cmd == 'INTERPOLATION':  # Texture1D, Texture2D, Texture3D
+                ob.set_interpolation(*args)
+            else:
+                logger.warning('Invalid GLIR command %r' % cmd)
+
```
Comment (Eric89GXL) [inline vispy/gloo/glir.py]: This is all copy-pasted without modification, right?

**Comment by Eric89GXL** [comment]
Other than my clarification, LGTM!

File: vispy/gloo/glir.py (hunk 4 above)
Comment (izaid) [inline vispy/gloo/glir.py]: Yep.

**Comment by izaid** [comment]
@rossant Except for tests, this is ready to go. How exactly should I do tests for this?

File: vispy/gloo/context.py [hunk 5]
```diff
@@ -160,7 +161,13 @@ class GLShared(object):
...
     
     def __init__(self):
-        self._parser = GlirParser()
+        glir_file = config['glir_file']
+
+        parser_cls = GlirParser
+        if glir_file:
+            parser_cls = glir_logger(parser_cls, glir_file)
+
+        self._parser = parser_cls()
```
Comment (almarklein) [inline vispy/gloo/context.py]: Note the `parser.setter` a few lines below. I think these additions should go in there.

**Comment by almarklein** [comment]
@rossant this seems very much related to the glir export mechanism that you want. Or is that the idea and am I simply put off by the name logger?

**Comment by izaid** [comment]
@almarklein That's the idea, Cyrille and I discussed this. I called it logger.

**Comment by almarklein** [comment]
Ok :+1:

**Comment by izaid** [comment]
@almarklein With this technique, it can't go into parser.setter because we apply a decorator to the class. The parser.setter method takes an already constructed class.

**Comment by almarklein** [comment]
Does this mean that the logger is disabled if a parser is set at the context?

**Comment by izaid** [comment]
Yes. If you change the parser, the logger is disabled. The right way to handle this would be to change which parser is being created. (I think this makes sense.)

**Comment by almarklein** [comment]
> How exactly should I do tests for this?

You can setup a simple visualization (e.g. just a few glClear commands), log that to a file (`import tempfile`), and check that the output is correct.

**Comment by izaid** [comment]
@almarklein For testing, it's not clear to me how I can send Glir commands to the parser without setting up a canvas and everything. If I just send Glir commands to the parser without setting anything else up, I get a segfault.

What is the simplest infrastructure I need to create to get this to work? Is there an example somewhere?

**Comment by Eric89GXL** [comment]
Look at the existing gl tests, gloo tests, app tests, or visual tests, we
do this in a number of places. Most will involve @requires_application and
"with Canvas", others monkey patch gloo (I think? ) to avoid actually
sending calls. Probably better to use the requires_application way.

File: vispy/gloo/tests/test_glir.py [hunk 6]
```diff
@@ -1,8 +1,12 @@
...
+import tempfile
 
-from vispy.testing import run_tests_if_main
+from vispy import config
+from vispy.app import Canvas
+from vispy.gloo import glir
+from vispy.testing import requires_application, run_tests_if_main
 
 
 def test__queue():
```
Comment (rossant) [inline vispy/gloo/tests/test_glir.py]: @Eric89GXL @almarklein  is the double `_` intended?

File: vispy/gloo/tests/test_glir.py [hunk 7]
```diff
@@ -45,6 +49,34 @@ def test__queue():
...
+    assert lines[i] == json.dumps(['FUNC', 'glClearColor', 1.0, 1.0, 1.0, 1.0])
+    i += 1
+    assert lines[i] == json.dumps(['FUNC', 'glClear', 17664])
+    i += 1
+    assert lines[i] == json.dumps(['FUNC', 'glFinish'])
+    i += 1
+
+    config.update(glir_file='')
+
+
```
Comment (rossant) [inline vispy/gloo/tests/test_glir.py]: delete the file at the end?

Comment (izaid) [inline vispy/gloo/tests/test_glir.py]: It's auto-deleted by tempfile.

File: vispy/gloo/tests/test_glir.py (hunk 6 above)
Comment (almarklein) [inline vispy/gloo/tests/test_glir.py]: probably not

**Comment by izaid** [comment]
This is *nearly* ready to go. I've updated this to include a test. It also is passing on every build configuration except one of Travis.

I can't figure out exactly what the problem is for that Travis configuration. It looks like Python 3.4 is unhappy for some stupid, silly reason, but I can't determine why. (Python 3.4 works fine on my Mac.)

Can someone else give this a shot? Maybe someone that uses Linux?

**Comment by Eric89GXL** [comment]
The errors look related to encoding/decoding strings, and/or how files are opened. When I activate Python 3.4 environment on OSX I get simliar errors, e.g.:
```
  File "/Users/larsoner/custombuilds/vispy/vispy/gloo/glir.py", line 421, in _parse
    json.dump(command, self._file, cls=NumPyJSONEncoder)
  File "/Users/larsoner/anaconda/envs/py34/lib/python3.4/json/__init__.py", line 179, in dump
    fp.write(chunk)
TypeError: 'str' does not support the buffer interface
```
It's likely that you either need to use encode/decode, or open your file with the proper mode as binary or text.

**Comment by Eric89GXL** [comment]
...in this case at least one offending line can be changed to this:
```
glir_file = tempfile.TemporaryFile(mode='w')
```
Are you sure you're running Python 3.4? These errors shouldn't really depend on the platform...

**Comment by izaid** [comment]
@Eric89GXL Thanks! The mode='w' will fail the test, as it needs read access. I'm just trying with mode='r+' to see if that works (the default for tempfile is mode='r+b').

Yes, I was trying with Python 3.4. I have no idea why it works here.

**Comment by Eric89GXL** [comment]
`r+` should work, it was the binary `b` that should have broken it

**Comment by Eric89GXL** [comment]
`test_glir.py` now passes over here on 3.4 and 2.7, which bodes well for Travis :)

**Comment by Eric89GXL** [comment]
Good to go from your end once Travis is happy? Looks like you've addressed previous comments

**Comment by rossant** [comment]
I have a comment actually: would it be possible for the output file to be valid JSON? So to have a list of lists?

**Comment by izaid** [comment]
Yep, good to go when green.

**Comment by izaid** [comment]
@rossant Is it cool if I change that in my next PR? I'll try to implement the load and execute in WebGL, which requires exactly that.

**Comment by rossant** [comment]
Also: can you make UNIFORM lists instead of base64 (or switch to lists if there are less than 20 values or something)? It might easier to read and to write the JS side.

**Comment by rossant** [comment]
@izaid sure!

**Comment by Eric89GXL** [comment]
sounds good, I'll merge when green then

**Comment by izaid** [comment]
@rossant @Eric89GXL :+1:

**Comment by Eric89GXL** [comment]
Thanks @izaid!

**Pull request merged**